Spanner: update samples to follow samples guide -1

laljikanjareeya opened the pull request:
Towards #1136 

This PR is mostly contain the Sample that require in the test setup.
All samples doesn't have separate test case as it covered in test setup itself.
Kept the old samples but Removed Test Setup.
In future PR I will make sure that all the related samples contains in one PR.

Here is The samples List:
- [x] Add Column(spanner_add_column)
- [x] Add Commit Timestamp(spanner_add_timestamp_column)
- [x] Add Index(spanner_create_index)
- [x] Add Storing Index(spanner_create_storing_index)
- [x] Create Backup(spanner_create_backup)
- [x] Create Custom Database(spanner_create_custom_database)
- [x] Create Database(spanner_create_database)
- [x] Create Instance(spanner_create_instance)
- [x] Create Table With Datatypes(spanner_create_table_with_datatypes)
- [x] Delete Backup(spanner_delete_backup)
- [x] Delete Database(Missing region tag)
- [x] Insert Sample Data(spanner_insert_data)
- [x] Insert Struct Sample Data(spanner_write_data_for_struct_queries)
- [x] Write Datatypes Data(spanner_insert_datatypes_data)

skuruppu commented on spanner/api/Spanner.Samples/InsertStructSampleDataAsync.cs:
```diff
@@ -0,0 +1,60 @@
+            new Singer {SingerId = 7, FirstName = "Gabriel", LastName = "Wright"},
+            new Singer {SingerId = 8, FirstName = "Benjamin", LastName = "Martinez"},
+            new Singer {SingerId = 9, FirstName = "Hannah", LastName = "Harris"},
+            new Singer {SingerId = 11, FirstName = "Anthony", LastName = "Hunter"},
+            new Singer {SingerId = 12, FirstName = "Earlean", LastName = "Holland"},
```
I'm curious why these two addition rows were added and the reason for the non-consecutive primary key.

laljikanjareeya replied:
Need more data for test  the samples. I will make sure that it should be consecutive. Thanks!!

amanda-tarafa replied:
I've left similar comments in other samples, but I think we should be receiving the list of singers as a parameters, and initialize that on tests. In my opinion this just adds clutter to the sample for someone who should be familiar with .NET (knows how to initialize objects and lists) and just want to know how to insert something in Spanner.

amanda-tarafa replied:
Also, if we remove the initialization from all the insert samples we might only need one sample to show in docs.

laljikanjareeya replied:
We can add singers as parameter but still we need it in initialization for query the data.

amanda-tarafa replied:
If this sample is part of the Getting Started guide, then it can remain as is. Else data initialization should be done as part of running tests, either in the fixture or in individual tests.

skuruppu commented on spanner/api/Spanner.Samples/AddColumnAsync.cs:
```diff
@@ -0,0 +1,37 @@
+    public async Task AddColumnAsync(string projectId, string instanceId, string databaseId)
+    {
+        // Initialize request argument(s).
+        string connectionString = $"Data Source=projects/{projectId}/instances/{instanceId}/databases/{databaseId}";
```
Here and in the files below, there is a reason why this string was line-wrapped in Program.cs. It's because if the line is too long, you can't see it all when the sample loads on the page (e.g. on laptop screens) and then you have to scroll right to see it. I don't know if we have a limit specified somewhere but I would cut off line lengths to 100.

@amanda-tarafa let us know if you have any guidance on this.

laljikanjareeya commented on spanner/api/Spanner.Samples.Tests/CreateBackupTest.cs:
```diff
@@ -0,0 +1,35 @@
+    {
+        CreateBackupSample createBackupSample = new CreateBackupSample();
+        // Backup is already exists it should throw exception.
+        Assert.Throws<RpcException>(() => createBackupSample.CreateBackup(_spannerFixture.ProjectId, _spannerFixture.InstanceId, _spannerFixture.BackupDatabaseId, _spannerFixture.BackupId));
```
Because backup has been already created in test setup.

skuruppu replied:
Oh I see, thanks for the explanation.

laljikanjareeya commented on spanner/api/Spanner.Samples/CreateCustomDatabaseAsync.cs:
```diff
@@ -0,0 +1,35 @@
+        {
+            string createStatement = $"CREATE DATABASE `{databaseId}`";
+            var cmd = connection.CreateDdlCommand(createStatement);
+            await cmd.ExecuteNonQueryAsync();
```
we are not putting try/catch in sample any more WDYT @amanda-tarafa

laljikanjareeya replied:
In original sample it was checking database is already exists or not.

skuruppu replied:
The original catch is:

```
                catch (SpannerException e) when (e.ErrorCode == ErrorCode.AlreadyExists)
                {
                    // Database Already Exists.
                }
```

so it's not doing anything useful. I think it's ok to not include it in the new sample.

skuruppu commented on spanner/api/Spanner.Samples.Tests/CreateBackupTest.cs:
```diff
@@ -0,0 +1,35 @@
+    public void TestCreateBackup()
+    {
+        CreateBackupSample createBackupSample = new CreateBackupSample();
+        // Backup is already exists it should throw exception.
```
```suggestion
        // Backup already exists since it was created in the test setup so it should throw an exception.
```

amanda-tarafa commented:
> I just have some small questions and also would like @amanda-tarafa's thoughts on the setup.

I'll get to this one probably next week.

amanda-tarafa commented:
Haven't forgotten about this, I'll get to it on Monday.

amanda-tarafa commented on spanner/api/Spanner.Samples/InsertSampleDataAsync.cs:
```diff
@@ -0,0 +1,86 @@
+        };
+        List<Album> albums = new List<Album>
+        {
+            new Album { SingerId = firstSingerId, AlbumId = 1, AlbumTitle = "Total Junk" },
+            new Album { SingerId = firstSingerId, AlbumId = 2, AlbumTitle = "Go, Go, Go" },
+            new Album { SingerId = secondSingerId, AlbumId = 1, AlbumTitle = "Green" },
+            new Album { SingerId = secondSingerId, AlbumId = 2, AlbumTitle = "Forever Hold your Peace" },
+            new Album { SingerId = secondSingerId, AlbumId = 3, AlbumTitle = "Terrified" },
```
Just asking, can two different albums have the same AlbumId?

skuruppu replied:
The primary key for this table is (SingerId, AlbumId) so you can have non-unique AlubumIds. You wouldn't be able to have duplicate SingerIds because that's the primary key of the Singer table.

amanda-tarafa replied:
OK

amanda-tarafa commented on spanner/api/Spanner.Samples/InsertSampleDataAsync.cs:
```diff
@@ -0,0 +1,86 @@
+                { "FirstName", SpannerDbType.String },
+                { "LastName", SpannerDbType.String }
+            });
+            await Task.WhenAll(singers.Select(singer =>
```
This might not work, it will be flaky for certain. You are using the same command for everything and thus the same parameter collection, whenever you change a parameter value,  it changes it for "everything".
You might want to move the command creation inside the Task.WhenAll action as well.

amanda-tarafa commented on spanner/api/Spanner.Samples/InsertSampleDataAsync.cs:
```diff
@@ -0,0 +1,86 @@
+        const int firstSingerId = 1;
+        const int secondSingerId = 2;
+        string connectionString = $"Data Source=projects/{projectId}/instances/{instanceId}/databases/{databaseId}";
+        List<Singer> singers = new List<Singer>
```
Why are we defining the data here? If we already have the Singer, Album classes, couldn't we just take the list of singers and albums as parameters? And hopefully we can reduce this sample to jus inserting in one table.

amanda-tarafa replied:
This one seems to be one of the ones used in the Getting Started guides, so if that's the case, it's fine to leave as is.

amanda-tarafa commented on spanner/api/Spanner.Samples/InsertSampleDataAsync.cs:
```diff
@@ -0,0 +1,86 @@
+                    { "AlbumTitle", SpannerDbType.String }
+                });
+
+            await Task.WhenAll(albums.Select(album =>
```
Same comment as above. This will be flaky.

amanda-tarafa commented on spanner/api/Spanner.Samples/Performance.cs:
```diff
@@ -0,0 +1,24 @@
+
+// [START spanner_insert_data_with_timestamp_column]
+
+public class Performance
```
Same comment as with the Album class, I actually think it's better if we add them as inner classes of all the sample classes that need it.

amanda-tarafa commented on spanner/api/Spanner.Samples/Performance.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public int SingerId { get; set; }
+    public int VenueId { get; set; }
+    public System.DateTime EventDate { get; set; }
```
```suggestion
    public DateTime EventDate { get; set; }
```
And add using System;

laljikanjareeya replied:
We have not added any import in classes that has been used in more than one sample as it might be conflict on import.

amanda-tarafa replied:
Ahh, that's a good point. And another reason to move these to the sample files where they are being used.

amanda-tarafa commented on spanner/api/Spanner.Samples/Singer.cs:
```diff
@@ -0,0 +1,31 @@
+// [START spanner_query_with_parameter]
+// [START spanner_quickstart
+
+public class Singer
```
Same, let's consider moving this to the samples that need it as inner classes of the sample class.

amanda-tarafa commented on spanner/api/Spanner.Samples/Venue.cs:
```diff
@@ -0,0 +1,45 @@
+    public string VenueName { get; set; }
+    public byte[] VenueInfo { get; set; }
+    public int Capacity { get; set; }
+    public System.Collections.Generic.List<System.DateTime> AvailableDates { get; set; }
+    public System.DateTime LastContactDate { get; set; }
+    public System.DateTime LastUpdateTime { get; set; }
```
usings for all these.

amanda-tarafa commented on spanner/api/Spanner.Samples/Venue.cs:
```diff
@@ -0,0 +1,45 @@
+// [START spanner_query_with_timestamp_parameter]
+// [START spanner_insert_datatypes_data]
+
+public class Venue
```
Same as before, better if we can move this to the sample classes.

amanda-tarafa commented on spanner/api/Spanner.Samples/WriteDatatypesDataAsync.cs:
```diff
@@ -0,0 +1,98 @@
+using System.Text;
+using System.Threading.Tasks;
+
+public class WriteDatatypesDataAsyncSample
```
```suggestion
public class InsertDatatypesDataAsyncSample
```
And file and method names accordingly.

amanda-tarafa commented on spanner/api/Spanner.Samples/InsertSampleDataAsync.cs:
```diff
@@ -0,0 +1,86 @@
+using System.Linq;
+using System.Threading.Tasks;
+
+public class InsertSampleDataAsyncSample
```
```suggestion
public class InsertDataAsyncSample
```

amanda-tarafa commented on spanner/api/Spanner.Samples/WriteDatatypesDataAsync.cs:
```diff
@@ -0,0 +1,98 @@
+
+        var availableDates1 = new List<DateTime>();
+        availableDates1.InsertRange(0, new DateTime[] {
+                DateTime.Parse("2020-12-01"),
```
This can fail or parse to an unexpected result depending on the Culture set in the environment where it's run.

amanda-tarafa replied:
```suggestion
                new DateTime(2020, 12, 01, 16, 32, 0, DateTimeKind.Utc),
```
And same with the rest. Parsing literal strings is not the best way to create dates in C#

amanda-tarafa commented on spanner/api/Spanner.Samples/WriteDatatypesDataAsync.cs:
```diff
@@ -0,0 +1,98 @@
+        availableDates3.InsertRange(0, new DateTime[] {
+                DateTime.Parse("2020-10-01"),
+                DateTime.Parse("2020-10-07")});
+        List<Venue> venues = new List<Venue> {
```
I really think we should receive a IList<Venue> as parameter and insert that. And then, there's really nothing particular about this sample that would distinguish it from the InsertDataAsync sample, the library accepts the usual .NET types and "converts" them to the Spanner types. I'm fine if we need to show just that, i.e. there's nothing special to be done, but then I would remove all the data parsing and byte array initialization code. Users should know how to work with dates and byte arrays.
@skuruppu thoughts welcome.

skuruppu replied:
Hmmm I can't see this sample being used anywhere, but it is consistently implemented this way in other languages as well:
- [Ruby](https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/8ce9a8db19d5f77dfca263d7064947e51160a901/spanner/spanner_samples.rb#L1068)
- [Golang](https://github.com/GoogleCloudPlatform/golang-samples/blob/62070bfa82e1a23e87ee84d563a43deb4b386d7c/spanner/spanner_snippets/spanner/spanner_insert_datatypes_data.go#L17)

We can talk about these offline and maybe include the docs team to clarify if we should delete them or not.

amanda-tarafa replied:
Waiting for the docs team to confirm.

amanda-tarafa commented on spanner/api/Spanner.Samples/WriteDatatypesDataAsync.cs:
```diff
@@ -0,0 +1,98 @@
+                        {"PopularityScore", SpannerDbType.Float64},
+                        {"LastUpdateTime", SpannerDbType.Timestamp},
+            });
+            await Task.WhenAll(venues.Select(venue =>
```
As with previous samples, this will be flaky.

amanda-tarafa commented on spanner/api/Spanner.Samples/InsertStructSampleDataAsync.cs:
```diff
@@ -0,0 +1,60 @@
+using System.Linq;
+using System.Threading.Tasks;
+
+public class InsertStructSampleDataAsyncSample
```
```suggestion
public class InsertStructDataAsyncSample
```

amanda-tarafa commented on spanner/api/Spanner.Samples/InsertStructSampleDataAsync.cs:
```diff
@@ -0,0 +1,60 @@
+                        {"FirstName", SpannerDbType.String},
+                        {"LastName", SpannerDbType.String}
+            });
+            var rows = await Task.WhenAll(singers.Select(singer =>
```
Same as with other insert samples, this will be flaky.

laljikanjareeya commented on spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:
```diff
@@ -0,0 +1,250 @@
+        {
+            try
+            {
+                DeleteBackupSample deleteBackupSample = new DeleteBackupSample();
```
We are not creating backup every time as it timeout. so delete backup sample also don't have test case.

amanda-tarafa replied:
OK.

laljikanjareeya commented on spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:
```diff
@@ -0,0 +1,250 @@
+            // We intentionally keep an existing database around to reduce the
+            // the likelihood of test timeouts when creating a backup so
+            // it's ok to get an AlreadyExists error.
+            if (e.ToString().Contains("Database already exists"))
```
we don't have `RpcException` exception here as it creates using `CreateDdlCommand` command, I will add catch..(when)...

amanda-tarafa replied:
OK

laljikanjareeya commented on spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:
```diff
@@ -0,0 +1,250 @@
+    public void Dispose()
+    {
+
+        DeleteDatabaseAsyncSample deleteDatabaseAsyncSample = new DeleteDatabaseAsyncSample();
+        foreach (var databaseId in DatabasesToDelete)
+        {
+            try
+            {
+                deleteDatabaseAsyncSample.DeleteDatabaseAsync(ProjectId, InstanceId, databaseId).Wait();
+            }
+            catch (Exception) { }
+        }
```
Currently Setup is removing all the old databases and backups so you mean to remove this from here?

skuruppu replied:
I'm curious why this is needed if on line 76 we have "// Delete all the databases created before 4 hrs.".

amanda-tarafa replied:
Yes, exactly, let's just do it once, on startup, possibly with 72 hours instead of 4.

laljikanjareeya commented on spanner/api/Spanner.Samples/AddStoringIndexAsync.cs:
```diff
@@ -0,0 +1,37 @@
+    {
+        // Initialize request argument(s).
+        string connectionString = $"Data Source=projects/{projectId}/instances/{instanceId}/databases/{databaseId}";
+        string createStatement = "CREATE INDEX AlbumsByAlbumTitle2 ON Albums(AlbumTitle) STORING (MarketingBudget)";
```
`AlbumsByAlbumTitle` index is already been use. I will give proper name here.

skuruppu replied:
Note that it's called `AlbumsByAlbumTitle2` in the [docs](https://cloud.google.com/spanner/docs/getting-started/csharp#using_the_client_library_for). I know that we'll have to rewrite the docs after this work is complete but wanted to make a note of it here.

amanda-tarafa replied:
I'm fine with keeping as is if it's going to be a problem in docs. This is just a little ugly that's all.

skuruppu replied:
I agree. But they do explicitly refer to it this way in the docs :(
```
Now you can execute a read that fetches all `AlbumId`, `AlbumTitle`, and `MarketingBudget` columns from the `AlbumsByAlbumTitle2` index
```

laljikanjareeya commented on spanner/api/Spanner.Samples/Album.cs:
```diff
@@ -0,0 +1,49 @@
+// [START spanner_read_stale_data]
+// [START spanner_query_with_query_options]
+
+public class Album
```
If we add it on sample where it's been used it need private class and we wan't be able to test it properly as in some case we are returning the list of objects.

amanda-tarafa replied:
You don't need the private class if you add it as an inner class of the class that contains the sample, roughly:
```suggestion
public class WhereAlbumIsUsedSample
{
    public class Album {...}
    
    public int UseAlbumSample(...) {...}
}
```

laljikanjareeya commented on spanner/api/Spanner.Samples/DeleteBackup.cs:
```diff
@@ -0,0 +1,34 @@
+        BackupName backupAsBackupName = BackupName.FromProjectInstanceBackup(projectId, instanceId, backupId);
+        databaseAdminClient.DeleteBackup(backupAsBackupName);
+
+        Console.WriteLine("Backup deleted successfully.");
```
I don't think `DeleteBackup` is long running operation.

skuruppu replied:
No, it's not long-running so it will be deleted immediately.

amanda-tarafa replied:
OK

laljikanjareeya commented on spanner/api/Spanner.Samples/DeleteDatabaseAsync.cs:
```diff
@@ -0,0 +1,34 @@
+        {
+            await cmd.ExecuteNonQueryAsync().ConfigureAwait(false);
+        }
+        Console.WriteLine($"Done deleting database: {databaseId}");
```
It deletes the database.

skuruppu commented on spanner/api/Spanner.Samples/AddStoringIndexAsync.cs:
```diff
@@ -0,0 +1,37 @@
+    {
+        // Initialize request argument(s).
+        string connectionString = $"Data Source=projects/{projectId}/instances/{instanceId}/databases/{databaseId}";
+        string createStatement = "CREATE INDEX AlbumsByAlbumTitle2 ON Albums(AlbumTitle) STORING (MarketingBudget)";
```
Ah right, that's a good point. This sample is used in the [CSharp Getting Started Guide](https://cloud.google.com/spanner/docs/getting-started/csharp#using_the_client_library_for), which is a series of steps that you follow to get all this setup. So in that context, it makes sense. But I agree that as a stand-alone sample, it won't work.

We can in the future consider re-writing the getting started guides. But I don't think the docs team will have the cycles to get to this until Q1 2021 and even then, we'd have to put in the request to do so now.

Let me know if you're ok with leaving this sample as is.

amanda-tarafa replied:
We have discussed this offline, and the samples on the Getting Started guide will remain as they currently are.

skuruppu commented on spanner/api/Spanner.Samples/CreateDatabaseAsync.cs:
```diff
@@ -0,0 +1,62 @@
+        // Initialize request connection string for database creation.
+        string connectionString = $"Data Source=projects/{projectId}/instances/{instanceId}";
+        // Make the request.
+        using (var connection = new SpannerConnection(connectionString))
```
Hmm good point. I don't think CreateCustomDatabaseAsync is used anywhere. I can't find it in the docs.

amanda-tarafa replied:
We have chatted about this offline and there are some samples that are not being used in docs and that can probably be removed. We are waiting for confirmation from TW.

skuruppu replied:
I commented in the CreateCustomDatabaseAsync sample to say we can delete that one :)

skuruppu commented on spanner/api/Spanner.Samples/InsertSampleDataAsync.cs:
```diff
@@ -0,0 +1,86 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START spanner_insert_data]
```
This is the context for the sample in the [Getting Started Guide](https://cloud.google.com/spanner/docs/getting-started/csharp#write-data-with-mutations). I think once again, these are not stand-alone samples, they are used for a broader workflow of getting a bunch of tables setup so that you can do more complex operations later in the guide.

amanda-tarafa replied:
See previous comment. It's fine as it is.

skuruppu commented:
Thanks for the really detailed review @amanda-tarafa.

There are a couple of concerns I have with changing the content of the samples (the ones included in the docs):
1. The wording around the samples in docs pages are common for all languages. So for the sample to make sense in the context of the docs, they kind of have to be consistent across the client lib languages.
2. I think a decent proportion of the code samples are used in the [Getting Started Guides](https://cloud.google.com/spanner/docs/getting-started/csharp) which is a tutorial to highlight various Spanner capabilities. While each sample may look odd on its own, in the context of the guide, they need to have that content so that users can complete the tutorial. E.g. InsertSampleDataAsync (which is called WriteSampleDataAsync in this PR) needs to insert those specific Singers and Albums so that the output in the later samples is consistent with what's shown in the docs.

I know that this makes things complicated. @amanda-tarafa we can speak more about this offline.

amanda-tarafa commented on spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:
```diff
@@ -0,0 +1,250 @@
+            // It's ok backup has been in progress in another test cycle.
+            if (e.Message.Contains("maximum number of pending backups (1) for the database has been reached"))
+            {
+                Console.WriteLine($"Backup {BackupId} already in progress.");
```
OK

amanda-tarafa commented:
@laljikanjareeya I'll review again when we have heard from the docs teams, some of these may go away but still not certain which, so I don't want to review something that we'll end up deleting. Let me know if there's something specific that you want me to take a look at.

skuruppu commented on spanner/api/Spanner.Samples/AddCommitTimestampAsync.cs:
```diff
@@ -0,0 +1,37 @@
+﻿// Copyright 2020 Google Inc.
```
This sample is unused in docs so is safe to delete.

skuruppu commented on spanner/api/Spanner.Samples/CreateCustomDatabaseAsync.cs:
```diff
@@ -0,0 +1,40 @@
+﻿// Copyright 2020 Google Inc.
```
This sample is safe to delete (it's only in C#).

skuruppu commented on spanner/api/Spanner.Samples/CreateTableWithDatatypesAsync.cs:
```diff
@@ -0,0 +1,49 @@
+﻿// Copyright 2020 Google Inc.
```
This sample is not used anywhere so it's safe to delete.

skuruppu commented on spanner/api/Spanner.Samples/DeleteBackup.cs:
```diff
@@ -0,0 +1,34 @@
+        DatabaseAdminClient databaseAdminClient = DatabaseAdminClient.Create();
+
+        // Make the DeleteBackup request.
+        BackupName backupAsBackupName = BackupName.FromProjectInstanceBackup(projectId, instanceId, backupId);
```
```suggestion
        BackupName backupAsBackupName =
            BackupName.FromProjectInstanceBackup(projectId, instanceId, backupId);
```

skuruppu commented on spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs:
```diff
@@ -0,0 +1,114 @@
+﻿// Copyright 2020 Google Inc.
```
This sample is safe to delete.

skuruppu commented on spanner/api/Spanner.Samples/InsertStructDataAsync.cs:
```diff
@@ -0,0 +1,68 @@
+﻿// Copyright 2020 Google Inc.
```
This sample is safe to delete.

laljikanjareeya commented:
@amanda-tarafa Thanks for your time for review. I Addressed all the comment and removed unused samples based on @skuruppu's feedback, Please take a look.

amanda-tarafa commented on spanner/api/Spanner.Samples/CreateDatabaseAsync.cs:
```diff
@@ -0,0 +1,61 @@
+        {
+            // Define create table statement for table #1.
+            string createTableStatement =
+           @"CREATE TABLE Singers (
```
nit: One space missing?
```suggestion
            @"CREATE TABLE Singers (
```

amanda-tarafa commented on spanner/api/Spanner.Samples/CreateDatabaseAsync.cs:
```diff
@@ -0,0 +1,61 @@
+            // Define create table statement for table #1.
+            string createTableStatement =
+           @"CREATE TABLE Singers (
+                     SingerId INT64 NOT NULL,
+                     FirstName    STRING(1024),
+                     LastName STRING(1024),
+                     ComposerInfo   BYTES(MAX)
```
nit: The spacing here is all different.
```suggestion
                     SingerId INT64 NOT NULL,
                     FirstName STRING(1024),
                     LastName STRING(1024),
                     ComposerInfo BYTES(MAX)
```

amanda-tarafa commented on spanner/api/Spanner.Samples/CreateDatabaseAsync.cs:
```diff
@@ -0,0 +1,61 @@
+            // Define create table statement for table #2.
+            createTableStatement =
+            @"CREATE TABLE Albums (
+                     SingerId     INT64 NOT NULL,
+                     AlbumId      INT64 NOT NULL,
+                     AlbumTitle   STRING(MAX)
```
Maybe don't align here, or aling on both:
```suggestion
                     SingerId INT64 NOT NULL,
                     AlbumId INT64 NOT NULL,
                     AlbumTitle STRING(MAX)
```

amanda-tarafa commented on spanner/api/Spanner.Samples/CreateBackup.cs:
```diff
@@ -0,0 +1,63 @@
+            DatabaseAsDatabaseName = DatabaseName.FromProjectInstanceDatabase(projectId, instanceId, databaseId),
+            ExpireTime = DateTime.UtcNow.AddDays(14).ToTimestamp()
+        };
+        InstanceName parentAsInstanceName = InstanceName.FromProjectInstance(projectId, instanceId);
+
+        // Make the CreateBackup request.
+        Operation<Backup, CreateBackupMetadata> response = databaseAdminClient.CreateBackup(parentAsInstanceName, backup, backupId);
```
```suggestion
        InstanceName instanceName = InstanceName.FromProjectInstance(projectId, instanceId);

        // Make the CreateBackup request.
        Operation<Backup, CreateBackupMetadata> response = databaseAdminClient.CreateBackup(instanceName, backup, backupId);
```

amanda-tarafa commented on spanner/api/Spanner.Samples.Tests/runTests.ps1:
```diff
@@ -12,13 +12,6 @@
 # the License.
 
-# TODO: Resurrect this test when bug 68199801 is fixed.
 Import-Module ..\..\..\BuildTools.psm1 -DisableNameChecking
```
We don't need this either.

amanda-tarafa commented on spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:
```diff
@@ -0,0 +1,250 @@
+        }
+        catch (RpcException ex) when (ex.Status.StatusCode == StatusCode.NotFound)
+        {
+            CreateInstanceSample createInstanceSample = new CreateInstanceSample();
```
Ping on this.

amanda-tarafa replied:
Ping on this again, happy to defer to another PR, or if you think it's really a bad idea to have a test for create instance because of timeouts, etc., let me know. Else, I still think it's a good idea to have it, this line of code will almost never run in CI I think.

amanda-tarafa replied:
Even if you just catch the "Instance already exists exception" the same way we have for CreateBackup, that's fine, it means the sample is working as intended.

amanda-tarafa commented on spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:
```diff
@@ -0,0 +1,229 @@
+
+        // Delete all the databases created before 72 hrs.
+        var timestamp = DateTimeOffset.UtcNow.AddHours(-72).ToUnixTimeMilliseconds();
+        foreach (var database in databases)
+        {
+            var databaseId = database.DatabaseName.DatabaseId.Replace("my-restore-db-", "").Replace("my-db-", "");
+            if (long.TryParse(databaseId, out long dbCreationTime) && dbCreationTime <= timestamp)
+            {
+                databasesToDelete.Add(database.DatabaseName.DatabaseId);
+            }
+        }
```
I want to insist on this: I know that `createTime` says "if exists", but are we certain that it's actually empty sometimes?
@skuruppu Can you confirm with the API team?

It would be so much cleaner to add that condition to the Where a few lines above instead of having to do this.

skuruppu replied:
Sorry for the delay here. I've sent a question to the API team and I'm awaiting a reply.

skuruppu replied:
I've got the answer. The `createTime` is currently not set on databases that weren't created as part of a RestoreDatabase request meaning that a database created with DatabaseCreate will not have this field set. But in an upcoming release, this field will be set for all databases.

I'm not sure what the timeline is but I think it should take about a month.

amanda-tarafa replied:
OK then, we can revisit later.

amanda-tarafa commented on spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:
```diff
@@ -0,0 +1,229 @@
+    private async Task DeleteDatabaseAsync(string databaseId)
+    {
+        string adminConnectionString = $"Data Source=projects/{ProjectId}/instances/{InstanceId}";
+        using (var connection = new SpannerConnection(adminConnectionString))
+        {
+            using (var cmd = connection.CreateDdlCommand($@"DROP DATABASE {databaseId}"))
+            {
+                await cmd.ExecuteNonQueryAsync();
+            }
+        }
```
nit: since C#8 you can do as follows. The using scope becomes the same as the declaration scope. Maybe consider doing this in the samples as well.
```suggestion
        using var connection = new SpannerConnection(adminConnectionString)
        using var cmd = connection.CreateDdlCommand($@"DROP DATABASE {databaseId}")
        await cmd.ExecuteNonQueryAsync();
```

amanda-tarafa reviewed (commented):
Mostly nits. This is looking great!

skuruppu commented on spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:
```diff
@@ -0,0 +1,223 @@
+            Console.WriteLine($"Backup {BackupId} already exists.");
+        }
+        catch (RpcException e) when (e.StatusCode == StatusCode.FailedPrecondition
+        && e.Message.Contains("maximum number of pending backups (1) for the database has been reached"))
```
Does this need to be indented?

laljikanjareeya replied:
Yes, As we are not deleting the backup that has been created in test setup because of timeout issue.
When setup runs first time or backup has been expire, it checks if backup has been already in progress on other environment or setup.

skuruppu commented on spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:
```diff
@@ -0,0 +1,223 @@
+        {
+            var cmd = connection.CreateUpdateCommand("Albums", new SpannerParameterCollection
+                {
+                    {"SingerId", SpannerDbType.Int64},
+                    {"AlbumId", SpannerDbType.Int64},
+                    {"MarketingBudget", SpannerDbType.Int64},
```
```suggestion
                    { "SingerId", SpannerDbType.Int64 },
                    { "AlbumId", SpannerDbType.Int64 },
                    { "MarketingBudget", SpannerDbType.Int64 },
```

skuruppu commented on spanner/api/Spanner.Samples/CreateDatabaseAsync.cs:
```diff
@@ -0,0 +1,61 @@
+            string createStatement = $"CREATE DATABASE `{databaseId}`";
+            var cmd = connection.CreateDdlCommand(createStatement);
+            await cmd.ExecuteNonQueryAsync();
+        }
```
nit

```suggestion
        }

```

skuruppu reviewed (approved):
Looking great @laljikanjareeya. I will followup on the create_time of Database questions. But looks good aside from a couple of small nits.

laljikanjareeya commented:
@amanda-tarafa @skuruppu If everything fine than we can go ahead for merge.
The question about `create_time` of Database can be fix in further PR.

amanda-tarafa reviewed (approved):
LGTM!
(I'll create two issues for what's pending here, so, happy to merge as is)
